File: driver-api/src/infrastructure/rabbitmq/rabbitmq.service.ts
import { Injectable, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { DriverAssignment, AssignmentStatus } from '../../domain/entities/driver-assignment.entity';
import * as amqp from 'amqplib';

@Injectable()
export class RabbitMQService implements OnModuleInit {
    constructor(
        private readonly configService: ConfigService,
        @InjectRepository(DriverAssignment)
        private readonly driverAssignmentRepository: Repository<DriverAssignment>,
    ) { }

    async onModuleInit() {
        console.log('✅ RabbitMQ service initialized (placeholder)');
        this.subscribeToShipmentAssigned();
    }

    async onModuleDestroy() {
        console.log('✅ RabbitMQ service destroyed');
    }

    async publishEvent(eventType: string, eventData: any, routingKey?: string): Promise<void> {
        const RABBITMQ_URL = process.env.RABBITMQ_URL || 'amqp://admin:password@rabbitmq';
        const EXCHANGE = 'logistics';
        const ROUTING_KEY = routingKey || eventType;

        try {
            const connection = await amqp.connect(RABBITMQ_URL);
            const channel = await connection.createChannel();

            // Declare exchange
            await channel.assertExchange(EXCHANGE, 'topic', { durable: true });

            // Publish message
            const message = JSON.stringify({
                eventType,
                data: eventData,
                timestamp: new Date().toISOString()
            });

            await channel.publish(EXCHANGE, ROUTING_KEY, Buffer.from(message));
            console.log(`📤 Event published: ${eventType} -> ${ROUTING_KEY}`, eventData);

            await channel.close();
            await connection.close();
        } catch (error) {
            console.error(`❌ Failed to publish event ${eventType}:`, error);
        }
    }

    async consumeEvents(queueName: string, callback: (message: any) => Promise<void>): Promise<void> {
        console.log(`👂 Consumer registered for queue: ${queueName}`);
    }

    async bindQueueToExchange(queueName: string, exchange: string, routingKey: string): Promise<void> {
        console.log(`🔗 Queue bound: ${queueName} to ${exchange} with key: ${routingKey}`);
    }

    async getQueueInfo(queueName: string): Promise<any> {
        return { queue: queueName, messageCount: 0 };
    }

    async purgeQueue(queueName: string): Promise<void> {
        console.log(`🧹 Queue purged: ${queueName}`);
    }

    async healthCheck(): Promise<boolean> {
        return true;
    }

    async subscribeToShipmentAssigned() {
        const RABBITMQ_URL = process.env.RABBITMQ_URL || 'amqp://admin:password@rabbitmq';
        const EXCHANGE = 'logistics';
        const ROUTING_KEY = 'shipment.assigned';
        const QUEUE = 'driver-api.shipment.assigned';

        try {
            const connection = await amqp.connect(RABBITMQ_URL);
            const channel = await connection.createChannel();

            // Declare exchange
            await channel.assertExchange(EXCHANGE, 'topic', { durable: true });

            // Declare queue
            await channel.assertQueue(QUEUE, { durable: true });

            // Bind queue to exchange with routing key
            await channel.bindQueue(QUEUE, EXCHANGE, ROUTING_KEY);

            channel.consume(QUEUE, async (msg) => {
                if (msg) {
                    try {
                        const event = JSON.parse(msg.content.toString());
                        console.log(`📨 Received shipment.assigned event:`, event);

                        const assignment = this.driverAssignmentRepository.create({
                            driverId: event.driverId,
                            shipmentId: event.shipmentId,
                            status: AssignmentStatus.PENDING,
                            assignedAt: event.assignedAt ? new Date(event.assignedAt) : new Date(),
                        });
                        await this.driverAssignmentRepository.save(assignment);
                        console.log(`✅ Assignment created for driver ${event.driverId} and shipment ${event.shipmentId}`);
                        channel.ack(msg);
                    } catch (err) {
                        console.error('❌ Error processing shipment.assigned event:', err);
                        channel.nack(msg, false, false);
                    }
                }
            });
            console.log('🚚 Listening for shipment.assigned events from RabbitMQ...');
        } catch (err) {
            console.error('❌ Failed to connect to RabbitMQ for shipment.assigned:', err);
        }
    }
} 

File: ml-route-optimizer/src/infrastructure/rabbitmq/rabbitmq_service.py
import pika
import json
import logging
import os
from typing import Callable, Dict, Any
from datetime import datetime

logger = logging.getLogger(__name__)

class RabbitMQService:
    """RabbitMQ service for ML Route Optimizer"""
    
    def __init__(self):
        self.connection = None
        self.channel = None
        self.rabbitmq_url = os.getenv('RABBITMQ_URL', 'amqp://admin:password@rabbitmq:5672')
        self.exchange = 'logistics'
        
    async def connect(self):
        """Connect to RabbitMQ"""
        try:
            self.connection = pika.BlockingConnection(
                pika.URLParameters(self.rabbitmq_url)
            )
            self.channel = self.connection.channel()
            
            # Declare exchange
            self.channel.exchange_declare(
                exchange=self.exchange,
                exchange_type='topic',
                durable=True
            )
            
            logger.info("✅ Connected to RabbitMQ")
            return True
        except Exception as e:
            logger.error(f"❌ Failed to connect to RabbitMQ: {str(e)}")
            return False
    
    async def publish_event(self, event_type: str, event_data: Dict[str, Any], routing_key: str = None):
        """Publish event to RabbitMQ"""
        try:
            if not self.channel:
                await self.connect()
            
            message = {
                'event_type': event_type,
                'data': event_data,
                'timestamp': datetime.utcnow().isoformat(),
                'source': 'ml-route-optimizer'
            }
            
            routing_key = routing_key or f"ml.{event_type}"
            
            self.channel.basic_publish(
                exchange=self.exchange,
                routing_key=routing_key,
                body=json.dumps(message),
                properties=pika.BasicProperties(
                    delivery_mode=2,  # make message persistent
                    content_type='application/json'
                )
            )
            
            logger.info(f"📤 Published event: {event_type} -> {routing_key}")
            
        except Exception as e:
            logger.error(f"❌ Failed to publish event: {str(e)}")
    
    async def subscribe_to_events(self, queue_name: str, routing_keys: list, callback: Callable):
        """Subscribe to events from RabbitMQ"""
        try:
            if not self.channel:
                await self.connect()
            
            # Declare queue
            self.channel.queue_declare(queue=queue_name, durable=True)
            
            # Bind queue to exchange with routing keys
            for routing_key in routing_keys:
                self.channel.queue_bind(
                    exchange=self.exchange,
                    queue=queue_name,
                    routing_key=routing_key
                )
            
            def message_handler(ch, method, properties, body):
                try:
                    message = json.loads(body.decode())
                    logger.info(f"📨 Received event: {message.get('event_type')}")
                    callback(message)
                    ch.basic_ack(delivery_tag=method.delivery_tag)
                except Exception as e:
                    logger.error(f"❌ Error processing message: {str(e)}")
                    ch.basic_nack(delivery_tag=method.delivery_tag, requeue=False)
            
            self.channel.basic_consume(
                queue=queue_name,
                on_message_callback=message_handler
            )
            
            logger.info(f"👂 Subscribed to events: {queue_name} -> {routing_keys}")
            
        except Exception as e:
            logger.error(f"❌ Failed to subscribe to events: {str(e)}")
    
    async def start_consuming(self):
        """Start consuming messages"""
        try:
            if self.channel:
                logger.info("🚀 Starting to consume messages...")
                self.channel.start_consuming()
        except Exception as e:
            logger.error(f"❌ Error consuming messages: {str(e)}")
    
    async def close(self):
        """Close RabbitMQ connection"""
        try:
            if self.connection:
                self.connection.close()
                logger.info("✅ RabbitMQ connection closed")
        except Exception as e:
            logger.error(f"❌ Error closing RabbitMQ connection: {str(e)}")
    
    async def health_check(self) -> bool:
        """Health check for RabbitMQ"""
        try:
            if not self.connection or self.connection.is_closed:
                return await self.connect()
            return True
        except Exception as e:
            logger.error(f"❌ RabbitMQ health check failed: {str(e)}")
            return False 